### PR TITLE
Kubeflow deployment name should default to the directory name.

### PR DESCRIPTION
**Pull request by jlewi**
* This matches 0.6 semantics by defaulting.

* More importantly this enables a user to do

  mkdir ${KFAPP}
  kfctl apply -V -f https://...

* Related to kubeflow/kfctl#52

* #4275 needs to be merged first.

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/kubeflow/4287)

### CODE REVIEW

**Comment by jlewi** [comment]
Failure is kubeflow/testing#480 ; out of address quota.

File: bootstrap/pkg/kfapp/coordinator/coordinator.go [hunk 1]
```diff
@@ -697,6 +697,23 @@ func LoadKfAppCfgFile(cfgfile string) (kftypesv3.KfApp, error) {
...
+		if err != nil {
+			return nil, &kfapis.KfError{
+				Code:    int(kfapis.INTERNAL_ERROR),
+				Message: fmt.Sprintf("KfDef.Name isn't set and there was a problem inferring the name based on the path %v; error: %v\nPlease set the name explicitly in the KFDef spec.", appFile, err),
+			}
+		}
+
+		appDir := filepath.Dir(absAppPath)
+
+		kfdef.Name = filepath.Base(appDir)
```
Comment (richardsliu) [inline bootstrap/pkg/kfapp/coordinator/coordinator.go]: If the app file is located directly under root (e.g. "~/app.yaml") this means the kfdef name is now "~". Will this cause problems?

Comment (jlewi) [reply bootstrap/pkg/kfapp/coordinator/coordinator.go]: Good catch. I correctly handle that now. If the name ends up being "" we throw and error.
PTAL

**Comment by jlewi** [comment]
Basic auth failed during delete

```
```

**Comment by jlewi** [comment]
/test all

Test flake looks like a kfctl_is_ready test flake waiting for Argo UI pods

**Comment by jlewi** [comment]
There were failures in delete

```
util.py                     71 INFO     Error: couldn't delete KfApp:  (kubeflow.error): Code 500 with message: coordinator Delete failed for gcp:  (kube
flow.error): Code 400 with message: Error when cleaning IAM policy: googleapi: Error 409: There were concurrent policy changes. Please retry the whole re
ad-modify-write with exponential backoff., aborted
```

**Comment by jlewi** [comment]
For basic auth the kfctl_is_ready_test failed.

File: bootstrap/pkg/kfapp/coordinator/coordinator.go [hunk 2]
```diff
@@ -335,6 +357,18 @@ func LoadKfAppCfgFile(cfgfile string) (kftypesv3.KfApp, error) {
 		}
 	}
 
+	// Since we know we have a local file we can set a default name if none is set based on the local directory
+	if kfdef.Name == "" {
+		kfdef.Name = nameFromAppFile(appFile)
+		if kfdef.Name == "" {
+			return nil, &kfapis.KfError{
+				Code:    int(kfapis.INTERNAL_ERROR),
```
Comment (gabrielwen) [inline bootstrap/pkg/kfapp/coordinator/coordinator.go]: shouldn't this be `INVALID_ARGUMENT`?  from the codes there are 2 error cases:
- not able to fetch absolute path.
- appFile lives in root directory.

seems to me both of the cases are error use of the CLI?  also root cause is `Spec.Name` is not set and we have to infer from the path, which is also on users?

File: bootstrap/pkg/kfapp/coordinator/coordinator.go [hunk 3]
```diff
@@ -280,6 +280,28 @@ func CreateKfAppCfgFile(d *kfdefsv3.KfDef) (string, error) {
...
 }
 
+// nameFromAppFile infers a default name given the path to the KFDef file.
+// returns the empty string if there is a problem getting the name.
+func nameFromAppFile(appFile string) string {
+	absAppPath, err := filepath.Abs(appFile)
+
+	if err != nil {
+		log.Errorf("KfDef.Name isn't set and there was a problem inferring the name based on the path %v; error: %v\nPlease set the name explicitly in the KFDef spec.", appFile, err)
+		return ""
```
Comment (gabrielwen) [inline bootstrap/pkg/kfapp/coordinator/coordinator.go]: shouldn't we pass the errors back instead of using empty string as an indication of errors?

Comment (jlewi) [reply bootstrap/pkg/kfapp/coordinator/coordinator.go]: I think its ok to use empty string. empty string indicates we couldn't get the default value. Passing  an error back to the caller code doesn't really help; it just forces the caller code to check the error. 

The primary reason I made this a separate function was just to write a unittest. 

The caller code provides an appropriate user facing error message in the event it ends up being the empty string.

**Comment by jlewi** [comment]
Most recent failure in delete was

```
2019-10-15T19:56:03.125955009Z  main            Traceback (most recent call last):
                                                  File "/usr/local/lib/python3.5/dist-packages/googleapiclient/discovery_cache/__init__.py", line 41, in autodetect
                                                    from . import file_cache
                                                  File "/usr/local/lib/python3.5/dist-packages/googleapiclient/discovery_cache/file_cache.py", line 41, in <module>
                                                    'file_cache is unavailable when using oauth2client >= 4.0.0')
                                                ImportError: file_cache is unavailable when using oauth2client >= 4.0.0

```

**Comment by jlewi** [comment]
Actually most recent failure was an issue deleting the endpoint.

```
2019-10-15T19:56:03.126054853Z  main            kfctl_delete_test.py        74 ERROR    Endpoint is not deleted: kfctl-9f56.endpoints.kubeflow-ci-deployment.cloud.goog
2019-10-15T19:56:03.126060626Z  main             generated xml file: /mnt/test-data-volume/kubeflow-presubmit-kfctl-go-iap-istio-4287-a9b59b5-9216-c198/output/artifacts/junit_ku
```

**Comment by jlewi** [comment]
It was a bug in the code; I wasn't actually invoking "run_delete" to delete kubeflow which is why verifying the endpoint no longer existed failed.

**Comment by jlewi** [comment]
Test failures were kubeflow/testing#480

/test all

**Comment by swiftdiaries** [comment]
Sorry I missed this
/lgtm

**Comment by jlewi** [comment]
Latest test failed on delete for basic auth.

```
util.py                     71 INFO     time="2019-10-15T23:59:24Z" level=info msg="Deleting kfctl-1aba status: RUNNING (op = operation-1571183110670-594
fb9330476d-d6c97af7-fb89d112)" filename="gcp/gcp.go:384"
util.py                     71 INFO     time="2019-10-15T23:59:45Z" level=info msg="Deleting kfctl-1aba status: RUNNING (op = operation-1571183110670-594
fb9330476d-d6c97af7-fb89d112)" filename="gcp/gcp.go:384"
util.py                     71 INFO     time="2019-10-16T00:00:06Z" level=info msg="Deleting kfctl-1aba status: RUNNING (op = operation-1571183110670-594
fb9330476d-d6c97af7-fb89d112)" filename="gcp/gcp.go:384"
util.py                     71 INFO     time="2019-10-16T00:00:47Z" level=info msg="Deleting kfctl-1aba status: RUNNING (op = operation-1571183110670-594
fb9330476d-d6c97af7-fb89d112)" filename="gcp/gcp.go:384"
util.py                     71 INFO     Error: couldn't delete KfApp:  (kubeflow.error): Code 500 with message: coordinator Delete failed for gcp:  (kube
flow.error): Code 400 with message: Gcp.Delete is failed for kubeflow-ci-deployment/kfctl-1aba: Deleting kfctl-1aba did not succeed; status: RUNNING (op
= operation-1571183110670-594fb9330476d-d6c97af7-fb89d112)
```

We ran 5 different versions of the E2E tests

* gcp_e2e_changed
* basic_auth
* go_build_deploy
* iap_istio

This could be too much.

**Comment by jlewi** [comment]
Here's the error for the delete operation test for kfctl-build-test 

```
gcloud --project=kubeflow-ci-deployment deployment-manager operations describe operation-1571183110670-594
fb9330476d-d6c97af7-fb89d112
ERROR: (gcloud.deployment-manager.operations.describe) ResponseError: code=404, message=The object 'projects/kubeflow-ci-deployment/global/operations/operation-1571183110670-594' is not found.
```

**Comment by jlewi** [comment]
For the basic auth test. The delete step failed

```
util.py                     71 INFO     Error: couldn't delete KfApp:  (kubeflow.error): Code 500 with message: coordinator Delete failed for gcp:  (kube
flow.error): Code 400 with message: Gcp.Delete is failed for kubeflow-ci-deployment/kfctl-4433: Deleting kfctl-4433 did not succeed; status: RUNNING (op
= operation-1571183346362-594fba13ca6bb-de0716e3-a9564c73)
```

But it looks like the operation eventually succeeded.

```
gcloud --project=kubeflow-ci-deployment deployment-manager operations describe operation-1571183346362-594fba13ca6bb-de0716e3-a9564c73
endTime: '2019-10-15T17:04:32.934-07:00'
id: '3079826449210532893'
insertTime: '2019-10-15T16:49:06.526-07:00'
kind: deploymentmanager#operation
name: operation-1571183346362-594fba13ca6bb-de0716e3-a9564c73
operationType: delete
progress: 100
selfLink: https://www.googleapis.com/deploymentmanager/v2/projects/kubeflow-ci-deployment/global/operations/operation-1571183346362-594fba13ca6bb-de0716e3-a9564c73
startTime: '2019-10-15T16:49:06.533-07:00'
status: DONE
targetId: '5636498988779432827'
targetLink: https://www.googleapis.com/deploymentmanager/v2/projects/kubeflow-ci-deployment/global/deployments/kfctl-4433
user: kubeflow-testing@kubeflow-ci.iam.gserviceaccount.com
```

It looks like it took 14 minutes for the operation to succeed.

**Comment by jlewi** [comment]
It doesn't seem like we are setting a max elapsed time so we end up using the default which is 15 minutes.

**Comment by jlewi** [comment]
I added retries to the test but that was only for 3 minutes so by the time `kubectl delete` error'd out the retry budget had been exhausted.

**Pull request merged**